poppy/wfe.py

```diff
@@ -353,6 +353,151 @@ def get_opd(self, wave):
         phase_screen = np.fft.ifftshift(np.fft.ifft2(np.fft.ifftshift(scaled))).real   # FT of scaled random PSD makes phase screen
 
         phase_screen -= np.mean(phase_screen)  # force zero-mean
-        opd = phase_screen / np.std(phase_screen) * self.wfe.to(u.m).value  # normalize to wanted input rms wfe
+        self.opd = phase_screen / np.std(phase_screen) * self.wfe.to(u.m).value  # normalize to wanted input rms wfe
 
-        return opd
+        return self.opd
+
+
+class PowerSpectrumWFE(WavefrontError):
+    """
+    Power spectrum PSD WFE class from characterizing and modeling 
+    optical surface power spectrum and applying optical noise.
+    
+    References:
+    Males, Jared. MagAO-X Preliminary-Design Review, 
+        Section 5.1: Optics Specifications, Eqn 1
+        https://magao-x.org/docs/handbook/appendices/pdr/
+    Lumbres, et al. In Prep.
+
+    Parameters
+    ----------
+    name : string
+        name of the optic
+    psd_parameters: list of array with various astropy quantities
+        Specifies the various PSD parameters with appropriate units, 
+        ordered according to which PSD model set. Follows the order:
+        [alpha, beta, outer_scale, inner_scale, surf_roughness]
+    psd_weight: iterable list of floats
+        Specifies the weight muliplier to set onto each model PSD
+    seed : integer
+        Seed for the random phase screen generator
+    apply_reflection: boolean
+        Applies 1/2 scale for the OPD as needed for reflection
+        Set to True if the PSD model does not account already for reflection.
+    screen_size: integer
+        Sets how large the PSD matrix will be calculated.
+        If None passed in, then code will default size to 4x wavefront's side.
+    wfe: astropy quantity
+        Optional. Use this to force the wfe RMS for opd surface.
+        If None passed, then the wfe RMS produced is what shows up in PSD calculation.
+    """
+
+    @utils.quantity_input(wfe=u.nm, radius=u.meter)
+    def __init__(self, name='Model PSD WFE', psd_parameters=None, psd_weight=None, seed=None, 
+                 apply_reflection=False, screen_size=None, wfe=None, **kwargs):
+
+        super().__init__(name=name, **kwargs)
+        self.psd_parameters = psd_parameters
+        self.seed = seed
+        self.apply_reflection = apply_reflection
+        self.screen_size = screen_size
+        
+        if hasattr(wfe, 'unit') or wfe is None:
+            self.wfe = wfe
+        else:
+            self.wfe = wfe * u.m # default assume meters
```

Comment:
This part should be unnecessary, given the `utils.quantity_input` decorator. That has the effect of ensuring that the supplied value of that parameter has the right unit (or an equivalent convertible one), or adding that unit if it's given a bare float. In this case, if the user supplies a float it will automatically get treated as nanometers and converted into `u.nm`, before even getting to this if/else statement.

Reply:
Okay. I have removed this block.